### ISSUE TITLE
Bundling code for libbpf does not work reliably.

### ISSUE THREAD

**Issue by Ferroin** [netdata/netdata#9570]
We're using `git apply` instead of just `patch` and utilizing a patch file stored in our source tree but not bundling that file into our dist archives.

We should either be using a custom fork of libbpf (we already select for a particular revision instead of pulling latest), or we should be properly bundling the patch and then applying it with `patch` instead of `git apply`

This causes the ebpf plugin install to fail on some systems.

**Comment by Ferroin** [comment]
@thiagoftsm @vlvkobal Your input here would be appreciated.

**Comment by vlvkobal** [comment]
If I remember well, patch is not available on all systems while git is a hard dependency for Netdata.

**Comment by vlvkobal** [comment]
> Git isn't available on all systems either

If you don't have `git` installed, you can't clone `libbpf` anyway, so what's the point of having `patch`?

> carrying our patch in our own fork

I considered the idea of having our own fork but I didn't like it.

**Comment by Ferroin** [comment]
> > Git isn't available on all systems either
> 
> If you don't have `git` installed, you can't clone `libbpf` anyway, so what's the point of having `patch`?

We're not cloning it, we're directly downloading a code archive from GitHub. And the advantage to using `patch` here is that it's a lot less picky about exact matches (IOW, it's more likely to work correctly if some code moves upstream).

> > carrying our patch in our own fork
> 
> I considered the idea of having our own fork but I didn't like it.

And did you consult with any of the other people who are going to be responsible for maintaining the installer code? Having our own fork with the required patches makes it much easier to handle in the installer, much less likely to break due to unforseen changes upstream, and a whole lot easier to diagnose if something goes wrong.

**Comment by vlvkobal** [comment]
> IOW, it's more likely to work correctly if some code moves upstream

How can it change if we are downloading one particular version and verify it's checksum?

**Comment by vlvkobal** [comment]
> We're not cloning it, we're directly downloading a code archive from GitHub

Ah, sorry - my mistake, I forgot about this. But anyway, `git` is installed by `install-required-packages.sh` and it is mentioned in our documentation as a required package. Does it make sense to add a new dependency?

**Comment by Ferroin** [comment]
> > IOW, it's more likely to work correctly if some code moves upstream
> 
> How can it change if we are downloading one particular version and verify it's checksum?

It _shouldn't_, but that doesn't mean that it _can't_. And when (yes, when, we're not sticking on the same version forever) we eventually update what version we're using, it's liely to require fewer (if any) changes to what we have if we're using `patch` instead of `git apply`.

> Does it make sense to add a new dependency?

I think it probably does make sense _because_ we're pulling in `git` already, because `git` usually pulls in `patch` on most systems. There are two advantages to not using `git apply` here, the reliability point I made above, and the fact that we can fetch release archives instead of downloading a packed copy of the repo (and thus have a smaller overall download size).

**Comment by Ferroin** [comment]
However, irrespective of all that, I still am strongly of the opinion that we should handle this like we handle libmosquitto and just carry the patch in a fork of the upstream repo and then download that directly so we don't have to patch code on a user's system (and thus don't run the risk of the patch applying incorrectly and causing other issues).

**Comment by Ferroin** [comment]
OK, I've verified now that #9490 _never_ worked for kickstart installs. I can't speak for @prologic, but I never realized it was even touching the install scripts, and even if I had I probably would not have thought to review it since it was opened while I was on vacation.

I can't see  I can't see any indication there either that the discussion around those changes actually involved @prologic, and I know it didn't involve me. It's very likely that had one of the two of us been involved, the implementation on the installer side would have been different, and we would not have an issue right now.

**Comment by prologic** [comment]
@Ferroin Its okay :) I'm not personally emotionally attached to any work I do! These things happen and IHMO it really wasn't a big deal here. I made sure I'm tagged as a code owner on most things to further ensure things don't accidentally slip through the cracks.

Anyway... What's the issue with the way this patch is tored/applied/etc? What's the problem at hand here? TL;DR please.

**Comment by Ferroin** [comment]
The big thing was that we did not include the patch in the dist archive.

As far as the `git apply` versus `patch` thing, that's more a practicality issue than anything else. We're currently downloading a copy of the repo as an archive and then using `git apply`, ut we should be just downloading the code at that revision (not the whole repo) and using `patch`, as it's more efficient.

Better yet though would be to just carry this as our own custom fork like we do for libmosquito (especially since we don't seem to be pushing these changes upstream), because then we just download a release that has the code already patched and don't need to rely on changes to code happening correctly on the user's system.

**Comment by vlvkobal** [comment]
> Better yet though would be to just carry this as our own custom fork

I'm starting to think maybe this way is a better one. It's easier to pull changes from upstream than maintaining an up to date patch. What about making releases? Does it need additional work to be done?

**Comment by Ferroin** [comment]
> > Better yet though would be to just carry this as our own custom fork
> 
> I'm starting to think maybe this way is a better one. It's easier to pull changes from upstream than maintaining an up to date patch. What about making releases? Does it need additional work to be done?

It should only need a small amount of work for each release (essentially, update things, then just create the release on GitHub).
